### PR TITLE
EZP-31413: Migrating a project to v3

### PR DESCRIPTION
**Pull request by DominikaK**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | https://jira.ez.no/browse/EZP-31413
| Versions      | all

### CODE REVIEW

File: docs/migrating/migrating_to_v3.md [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+    ALTER TABLE ezdatebasedpublisher_scheduled_version RENAME COLUMN publication_date TO date;
+    ALTER TABLE ezdatebasedpublisher_scheduled_version ADD COLUMN action VARCHAR(32);
+    UPDATE ezdatebasedpublisher_scheduled_version SET action = 'publish';
+    ALTER TABLE ezdatebasedpublisher_scheduled_version ALTER COLUMN action SET NOT NULL ;
+    COMMIT;
+    ```
+
+### Site Factory
+
+For production, it is recommended to create the `esite` and `ezsite_public_access` tables manually importing their schemas definition:
```
Comment (damianz5) [inline docs/migrating/migrating_to_v3.md:420]: `ezsite` instead of `esite`

File: docs/updating/upgrading_to_v3.md [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+
+### Update the database
+
+Apply the following database update script:
+
+`mysql -u <username> -p <password> <database_name> < vendor/ezsystems/ezpublish-kernel/blob/master/data/update/mysql/dbupdate-7.5.latest-to-8.0.0.sql`
+
+or for PostgreSQL:
+
+`psql <database_name> < vendor/ezsystems/ezpublish-kernel/blob/master/data/update/postgres/dbupdate-7.5.latest-to-8.0.0.sql`
```
Comment (mnocon) [inline docs/updating/upgrading_to_v3.md:447]: These links won't work, the scripts are now here: https://github.com/ezsystems/ezplatform/tree/master/upgrade/db

File: docs/updating/upgrading_to_v3.md [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+
+!!! enterprise
+
```
Comment (mnocon) [inline docs/updating/upgrading_to_v3.md:450]: And for Enterprise they are here: https://github.com/ezsystems/ezplatform-ee/tree/master/upgrade/db

so maybe it's not worth keeping them here?

File: docs/updating/upgrading_to_v3.md [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+    CREATE INDEX ezsite_public_access_site_id ON ezsite_public_access (site_id);
+    ```
+
+### Update Apache/Nginx configuration
+
+Make sure that you Apache/Nginx configuration is up to date with Symfony 5.
+Refer to [the provided `vhost.template`](https://github.com/ezsystems/ezplatform/blob/master/doc/apache2/vhost.template)
+for an example.
+
+## 5\. Update the database
```
Comment (mnocon) [inline docs/updating/upgrading_to_v3.md:552]: Now I see that this section: https://github.com/ezsystems/developer-documentation/pull/953/files?short_path=919ebf3#diff-877068fcb53101273f10ca25b5750738R439 should be probably removed?

File: docs/releases/ez_platform_v3.0_deprecations.md [hunk 5]
```diff
@@ -2,6 +2,11 @@
 
 This page lists backwards compatibility breaks and deprecations introduced in eZ Platform v3.0.
 
+!!! tip "Upgrade to v3"
+
+    For a guide on moving your project to v3,
+    see [Upgrading to v3](../updating/upgrading_to_v3.md).
```
Comment (MagdalenaZuba) [inline docs/releases/ez_platform_v3.0_deprecations.md:8]:
```suggestion
    see [Upgrading eZ Platform to v3](../updating/upgrading_to_v3.md).
```

File: docs/updating/upgrading_to_v3.md [hunk 6]
```diff
@@ -0,0 +1,560 @@
+# Upgrading to v3
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:1]:
```suggestion
# Upgrading eZ Platform to v3
```

File: docs/updating/upgrading_to_v3.md [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+
+To upgrade your v2 project to v3, you need to follow the [standard update procedure](../updating/updating_ez_platform.md),
+but you will also need to make changes to your code.
+
+## 1\. Automatically refactor your code
+
+To simplify the process of adapting your code to Symfony 5, you can use [rector](https://github.com/rectorphp/rector)
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:8]:
```suggestion
To simplify the process of adapting your code to Symfony 5, you can use [Rector, a reconstructor tool](https://github.com/rectorphp/rector)
```

File: docs/updating/upgrading_to_v3.md [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+to automatically refactor your Symfony and PHPunit code.
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:9]:
```suggestion
that will automatically refactor your Symfony and PHPunit code.
```

File: docs/updating/upgrading_to_v3.md [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+
+## 2\. Adapt project structure
+
+!!! tip
+
+    For details on all changes related to the switch to Symfony 5,
+    see [Symfony upgrade guide for 4.0](https://github.com/symfony/symfony/blob/master/UPGRADE-4.0.md)
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:16]:
```suggestion
    see [Symfony upgrade guide for 4.0](https://github.com/symfony/symfony/blob/5.0/UPGRADE-4.0.md)
```

File: docs/updating/upgrading_to_v3.md [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+    and [for 5.0](https://github.com/symfony/symfony/blob/master/UPGRADE-5.0.md)
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:17]:
```suggestion
    and [for 5.0](https://github.com/symfony/symfony/blob/5.0/UPGRADE-5.0.md)
```

File: docs/updating/upgrading_to_v3.md [hunk 11]
```diff
@@ -0,0 +1,560 @@
...
+The latest Symfony versions changed the organization of your project into folders and bundles.
+When updating to eZ Platform v3 you need to move your files and modify file paths and namespace references.
+
+![Project structure changes in v3](img/folder_structure_v3.png "Project folder structure changes between v2 and v3")
+
+#### Configuration
+
+Configuration files have been moved from `app/Resources/config` to `config`.
+Package-specific configuration is placed in `config/packages` (e.g. `config/packages/ezplatform_admin_ui.yaml`).
+This folder also contains `config/packages/ezplatform.yaml`, which contains all settings coming in from the Kernel.
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:28]:
```suggestion
This folder also contains `config/packages/ezplatform.yaml`, which contains all settings coming in from Kernel.
```

File: docs/updating/upgrading_to_v3.md [hunk 12]
```diff
@@ -0,0 +1,560 @@
...
+
+#### PHP code and bundle organization
+
+Since Symfony 4 your code is no longer organized in bundles.
+`AppBundle` has been removed from eZ Platform.
+You need to move all your PHP code, such as controllers or event listeners, in the `src` folder.
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:34]:
```suggestion
You need to move all your PHP code, such as controllers or event listeners, to the `src` folder.
```

File: docs/updating/upgrading_to_v3.md [hunk 13]
```diff
@@ -0,0 +1,560 @@
...
+Use the `App` namespace for your custom code instead.
+
+#### View templates
+
+Templates are no longer stored in `app/Resources/views`.
+You need to move all your templates in the `templates` folder in your project's root.
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:40]:
```suggestion
You need to move all your templates to the `templates` folder in your project's root.
```

File: docs/updating/upgrading_to_v3.md [hunk 14]
```diff
@@ -0,0 +1,560 @@
...
+
+#### `web` and assets
+
+Content of the `web` folder is now placed in `public`.
+Content of `app/Resources/assets` has been moved to `assets`.
+
+!!! note
+
+    You also need to update paths that refer to the old location,
+    for example in [`webpack.config.js`](../guide/bundles.md#configuration-from-a-bundle)
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:54]:
```suggestion
    for example in [`webpack.config.js`](../guide/bundles.md#configuration-from-a-bundle).
```

File: docs/updating/upgrading_to_v3.md [hunk 15]
```diff
@@ -0,0 +1,560 @@
...
+you now need to extend `EzSystemsPlatformHttpCache` instead, if you wish to make use of cache:
+`{% extends "@EzSystemsPlatformHttpCache/content_fields.html.twig" %}`.
+
+### Signal Slots and Events
+
+If you used Signal Slots to listen for events in you custom code,
+you need to rewrite them using Symfony Events and Listeners instead.
+
+The application now triggers two Events per operation: one before and one after the relevant thing happens
+(see for example [Bookmark events](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Event/BookmarkService.php)).
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:252]:
```suggestion
(see for example [Bookmark events](https://github.com/ezsystems/ezplatform-kernel/blob/v1.0.0/eZ/Publish/Core/Event/BookmarkService.php)).
```

File: docs/updating/upgrading_to_v3.md [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+
+To use them, create [Event Listeners](https://symfony.com/doc/4.3/event_dispatcher.html) in your code,
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:254]:
```suggestion
To use them, create [Event Listeners](https://symfony.com/doc/5.0/event_dispatcher.html) in your code,
```

File: docs/updating/upgrading_to_v3.md [hunk 17]
```diff
@@ -0,0 +1,560 @@
...
+### HTTP cache
+
+HTTP cache bundle now uses FOS Cache Bundle v2.
+If your code makes use of HTTP cache bundle, see [the list of changes and deprecations](../releases/ez_platform_v3.0_deprecations.md#http-cache-bundle).
+
+### REST
+
+If your code extends the REST API, you need to modify namespaces.
+The `eZ\Publish\Core\REST` and `eZ\Publish\Core\REST\Common\` namespaces have been replaced by `EzSystems\EzPlatformRest`.
+This is due to the fact that REST code has been moved from kernel to a new `ezpublish-rest` package.
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:334]:
```suggestion
This is due to the fact that REST code has been moved from Kernel to a new `ezpublish-rest` package.
```

File: docs/updating/upgrading_to_v3.md [hunk 18]
```diff
@@ -0,0 +1,560 @@
...
+because `ContainerAwareCommand` is deprecated.
+
+#### Permission-related methods
+
+Some deprecated [permission-related methods](../releases/ez_platform_v3.0_deprecations.md#permission-related-methods) have been removed.
+If your code uses them, you need to rewrite it to use the permission resolver.
+
+#### Container parameters
+
+A numer of Symfony Dependency Injection Container parameters [have been dropped](https://github.com/ezsystems/ezpublish-kernel/blob/master/doc/bc/8.0/dropped-container-parameters.md).
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:416]:
```suggestion
A numer of Symfony Dependency Injection Container parameters [have been dropped](https://github.com/ezsystems/ezplatform-kernel/blob/v1.0.0/doc/bc/1.0/dropped-container-parameters.md).
```

File: docs/updating/upgrading_to_v3.md [hunk 19]
```diff
@@ -0,0 +1,560 @@
...
+
+To check if your code uses them, search for all occurrences of `ezpublish\..*\.class` (regular expression pattern)
+and [decorate Symfony services](https://symfony.com/doc/4.4/service_container/service_decoration.html) instead.
```
Comment (MagdalenaZuba) [inline docs/updating/upgrading_to_v3.md:419]:
```suggestion
and [decorate Symfony services](https://symfony.com/doc/5.0/service_container/service_decoration.html) instead.
```

**Pull request merged**